kimchi/src/circuits/polynomials/foreign_field_mul/gadget.rs

```diff
@@ -0,0 +1,373 @@
+
+        for idx in 0..(cs.domain.d1.size() * 8) {
+            let table_id = match lcs.table_ids8.as_ref() {
+                Some(table_ids8) => table_ids8.evals[idx],
+                None =>
+                // If there is no `table_ids8` in the constraint system,
+                // every table ID is identically 0.
+                {
+                    F::zero()
+                }
+            };
+
+            let combined_entry = {
+                let table_row = lcs.lookup_table8.iter().map(|e| &e.evals[idx]);
+
+                lookup::tables::combine_table_entry(
+                    &joint_combiner,
+                    &table_id_combiner,
+                    table_row,
+                    &table_id,
+                )
+            };
+            evals.push(combined_entry);
+        }
+
+        Evaluations::from_vec_and_domain(evals, cs.domain.d8)
+    };
+
+    // Compute the sorted plookup evaluations
+    // TODO: Once we switch to committing using lagrange commitments, `witness` will be consumed when we interpolate,
+    //       so interpolation will have to moved below this.
+    let sorted: Vec<_> = lookup::constraints::sorted(
+        dummy_lookup_value,
+        &joint_lookup_table_d8,
+        cs.domain.d1,
+        &cs.gates,
+        witness,
+        joint_combiner,
+        table_id_combiner,
+        lookup_info,
+    )
+    .map_err(|_| CircuitGateError::InvalidLookupConstraintSorted(gate_type))?;
+
+    // Randomize the last `EVALS` rows in each of the sorted polynomials in order to add zero-knowledge to the protocol.
+    let sorted: Vec<_> = sorted
+        .into_iter()
+        .map(|chunk| lookup::constraints::zk_patch(chunk, cs.domain.d1, rng))
+        .collect();
+
+    let sorted_coeffs: Vec<_> = sorted.iter().map(|e| e.clone().interpolate()).collect();
+    let sorted8 = sorted_coeffs
+        .iter()
+        .map(|v| v.evaluate_over_domain_by_ref(cs.domain.d8))
+        .collect::<Vec<_>>();
+
+    // Compute the plookup aggregation evaluations
+    let aggreg = lookup::constraints::aggregation::<_, F>(
+        dummy_lookup_value,
+        &joint_lookup_table_d8,
+        cs.domain.d1,
+        &cs.gates,
+        witness,
+        &joint_combiner,
+        &table_id_combiner,
+        *beta,
+        *gamma,
+        &sorted,
+        rng,
+        lookup_info,
+    )
+    .map_err(|_| CircuitGateError::InvalidLookupConstraintAggregation(gate_type))?;
+
+    // Precompute different forms of the aggregation polynomial for later
+    let aggreg_coeffs = aggreg.interpolate();
+    // TODO: There's probably a clever way to expand the domain without interpolating
+    let aggreg8 = aggreg_coeffs.evaluate_over_domain_by_ref(cs.domain.d8);
+
+    Ok(LookupEnvironmentData {
+        aggreg8,
+        sorted8,
+        joint_lookup_table_d8,
+    })
+}
+
+// TODO: Check do we use this anywhere
+pub fn circuit_gate_selector_index(typ: GateType) -> usize {
```

Comment:
Just adding a note for the future to remember if we want to keep this approach for ALL gates, or maybe some other universal format for these, as otherwise it can look error-prone when used manually anywhere else in the codebase.

Reply:
this has been removed